runtime/src/main/java/org/corfudb/runtime/collections/StreamManager.java

```diff
@@ -0,0 +1,224 @@
+                                            CorfuStreamEntry.fromSMREntry(smrEntry,
+                                                    epoch,
+                                                    tablesOfInterest.get(streamId).getKeyClass(),
+                                                    tablesOfInterest.get(streamId).getPayloadClass(),
+                                                    tablesOfInterest.get(streamId).getMetadataClass())
+                                    ).collect(Collectors.toList())));
+
+                    if (!entries.isEmpty()) {
+                        CorfuStreamEntries callbackResult = new CorfuStreamEntries(entries);
+                        log.debug("{}::onNext with {} updates", listener.toString(), entries.size());
+                        long onNextStart = System.nanoTime();
+                        listener.onNext(callbackResult);
+                        long onNextEnd = System.nanoTime();
+                        if (TimeUnit.NANOSECONDS.toSeconds(onNextEnd - onNextStart) > 0) {
+                            log.debug("{}::onNext took {}s", listener.toString(),
+                                    TimeUnit.NANOSECONDS.toSeconds(onNextEnd - onNextStart));
+                        }
+                    }
+                }
+            } catch (Throwable throwable) {
+                log.warn("{}::onError: {}", listener.toString(), throwable.toString());
+                listener.onError(throwable);
+                streamManager.unsubscribe(listener);
+            }
+        }
+
+        public <K extends Message, V extends Message, M extends Message>
+        SubscriptionThread(@Nonnull StreamManager streamManager,
+                                  @Nonnull StreamListener listener,
+                                  @Nonnull String namespace,
+                                  @Nonnull List<TableSchema<K, V, M>> tableSchemas,
+                                  long startAddress) {
+            super(String.format("CorfuSubscriber: %s:%s", listener.toString(), namespace));
+            this.streamManager = streamManager;
+            this.listener = listener;
+            this.namespace = namespace;
+            this.tablesOfInterest = new HashMap<>();
+            this.shutdown = new AtomicBoolean();
+            tableSchemas.forEach(tableSchema -> tablesOfInterest.put(
+                    CorfuRuntime.getStreamID(
+                            TableRegistry.getFullyQualifiedTableName(namespace,
+                                    tableSchema.getTableName())
+                    ), tableSchema
+            ));
+
+            this.startAddress = startAddress;
+            StreamOptions options = StreamOptions.builder()
+                    .cacheEntries(false)
+                    .build();
+
+            this.txnStream = runtime.getStreamsView()
+                    .get(ObjectsView.TRANSACTION_STREAM_ID, options);
+        }
```

Comment:
On unsubscribe the stream is not removed from the StreamsView. A series of subscribe/unsubscribe can create a stream leak.

Reply:
thanks will do this.